### PR TITLE
browser(firefox): Page.setComposition

### PR DESCRIPTION
**Pull request by trueadm**
Extracted the Firefox browser patches changes from https://github.com/microsoft/playwright/pull/9535.

### CODE REVIEW

File: browser_patches/firefox-beta/BUILD_NUMBER [hunk 1]
```diff
@@ -1,2 +1,2 @@
-1300
-Changed: yurys@chromium.org Wed Oct 27 17:58:07 PDT 2021
+1301
+Changed: dg@domgan.com Mon 1 Nov 2021 13:04:12 GMT
```
Comment (mxschmitt) [inline browser_patches/firefox-beta/BUILD_NUMBER:2]: looks like the line ending has changed here (we usually use the script to update the file (browser_patches/export.sh).

Comment (mxschmitt) [inline browser_patches/firefox-beta/BUILD_NUMBER:2]: looks like the line ending has changed here (we usually use the script to update the file (browser_patches/export.sh).

File: browser_patches/firefox/BUILD_NUMBER [hunk 2]
```diff
@@ -1,2 +1,2 @@
-1298
-Changed: yurys@chromium.org Wed Oct 27 17:58:07 PDT 2021
+1299
+Changed: dg@domgan.com Mon 1 Nov 2021 13:04:12 GMT
```
Comment (mxschmitt) [inline browser_patches/firefox/BUILD_NUMBER:2]: ditto (line ending)

File: browser_patches/firefox/juggler/protocol/Protocol.js [hunk 3]
```diff
@@ -2,13 +2,15 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
-const {t, checkScheme} = ChromeUtils.import('chrome://juggler/content/protocol/PrimitiveTypes.js');
+const { t, checkScheme } = ChromeUtils.import(
```
Comment (mxschmitt) [inline browser_patches/firefox/juggler/protocol/Protocol.js:5]: looks like a lot of unrelated formatting changes did happen here, can we only include actual impl changes?

File: browser_patches/firefox/BUILD_NUMBER [hunk 4]
```diff
@@ -1,2 +1,2 @@
-1298
-Changed: yurys@chromium.org Wed Oct 27 17:58:07 PDT 2021
+1299
```
Comment (mxschmitt) [inline browser_patches/firefox/BUILD_NUMBER:1]: you need to bump the firefox-beta build number as well. Otherwise only firefox will trigger a new build.

File: packages/playwright-test/src/reporters/html.ts [hunk 5]
```diff
@@ -128,14 +128,14 @@ class HtmlReporter {
     const reportFolder = htmlReportFolder(this._outputFolder);
     await removeFolders([reportFolder]);
     const builder = new HtmlBuilder(reportFolder, this.config.rootDir);
-    const ok = builder.build(reports);
+    const { ok, singleTestId } = builder.build(reports);
```
Comment (mxschmitt) [inline packages/playwright-test/src/reporters/html.ts:131]: revert the changes in this file please.

Comment (trueadm) [reply packages/playwright-test/src/reporters/html.ts]: I messed up a rebase, sorry. This change wasn't intentional :)

Comment (trueadm) [reply packages/playwright-test/src/reporters/html.ts]: I messed up a rebase, sorry. This change wasn't intentional :)

**Review by mxschmitt — APPROVED** [review]

File: browser_patches/firefox-beta/juggler/content/PageAgent.js [hunk 6]
```diff
@@ -786,6 +787,37 @@ class PageAgent {
...
+      if (activeElement !== null) {
+        if (
+          activeElement.nodeName === "TEXTAREA" ||
+          activeElement.nodeName === "INPUT"
+        ) {
+          activeElement.setSelectionRange(replacementStart, replacementEnd);
+        } else {
+          const selection = window.getSelection();
+          const anchorNode = selection.anchorNode;
+          selection.setBaseAndExtent(anchorNode, replacementStart, anchorNode, replacementEnd);
```
Comment (JoelEinbinder) [inline browser_patches/firefox-beta/juggler/content/PageAgent.js:812]: This doesn't look quite right. Aside from the shadow dom issues, the macOS api this is modeled after expects this range to be in terms of the current root editable node. However your code is in terms of the leaf editable node.

You can see this code in TextInputHandler.mm : IMEInputHandler::SetMarkedText, which ends up in the web process as ContentEventHandler::OnSelectionEvent. I am working on a patch to fix this but I'm not sure if I should duplicate the logic in JavaScript or add a hack for us to call the c++ code directly.

**Review by JoelEinbinder — CHANGES_REQUESTED** [review]

**Review by raghavthind2005 — APPROVED** [review]

**Comment by pavelfeldman** [comment]
Closing due to no progress (on our side). I'm not sure what the action for this should be, but we aren't considering adding composition API at this point.
